### PR TITLE
[1] Add log file rotation functionality

### PR DESCRIPTION
**Pull request by jacobminer**
### Summary of Problem:
Android didn't support the log file rotation like iOS

### Proposed Solution:
Add log rotation functionality. Note that this doesn't work exactly the same as iOS because we don't have a method for accessing the creation date of a file. Instead, we used the last-modified date to determine if the file should be rotated.

With the default value of 10 minutes, this will only differ from iOS in the case where the app is open and is logging for a full 10 minutes. At that point, iOS would rotate the log file, where as Android wouldn't rotate until 10 after the last write. In both cases, the most likely scenario for a rotation is in between app launches though.

The rules for selecting a log file on Android are:
If any of the log files was last edited within the configured expiry time, use that file
Otherwise, use the first empty log file
If no empty or unexpired log file exists, clear the oldest one and use that instead

I've also implemented some caching to try to prevent performance issues caused by finding the correct log file each time.

### Testing Steps:
1. Clear all log files on the device using Android Studio's Device File Explorer (for my emulator, the files were in `/storage/emulated/0/Android/data/com.steamclock.steamclogsample/cache/logs`
2. Set the rotation time before logging anything to a low value (I used 5 seconds) in code
3. Wait X number of seconds (where X is your log rotation time) then tap the "Log Things" button in the app
4. Repeat the previous step 5 times
5. Ensure you have the correct number of log files in your log directory (should be one for the app starting, then one more for each "Log Things" button tap)
7. Repeat steps 3 and 4 again.
8. Ensure you have log files 0 - 9, and that the log files were correctly rotated based on the max of 10 files (ie: log file 11 should be log file 1, so it should have later logs than log file 5)
9. Tap "Show File Dump" and ensure logs are printed out in the correct order (by log time, and not file index)

### Screenshots:
N/A

### CODE REVIEW

File: steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt [hunk 1]
```diff
@@ -105,9 +104,13 @@ internal class ConsoleDestination: Timber.DebugTree() {
  */
 internal class ExternalLogFileDestination : Timber.DebugTree() {
     private var fileNamePrefix: String = "sclog"
-    private var fileNameTimestamp = "yyyy_MM_dd"
     private var fileExt = "txt"
 
+    private var rotatingIndexes = (0 until 10).map { it }
```
Comment (ssawchenko) [inline steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:109]: Not sure if it's any better or worse, but you could also initialize a List with an initializer function instead of running through a map:
`var rotatingIndexes = List(10) { it }`

This should also give you the List: `[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]`

File: steamclog/src/main/java/com/steamclock/steamclog/Steamclog.kt [hunk 2]
```diff
@@ -55,9 +55,10 @@ object SteamcLog {
     }
 
     fun initWith(isDebug: Boolean,
-                 fileWritePath: File? = null) {
+                 fileWritePath: File? = null,
+                 fileRotationSeconds: Long? = null) {
         fileWritePath?.let {
-            this.config = Config(isDebug, fileWritePath)
+            this.config = Config(isDebug, fileWritePath, autoRotateConfig = AutoRotateConfig(fileRotationSeconds ?: AutoRotateConfig.defaultFileRotationSeconds))
```
Comment (ssawchenko) [inline steamclog/src/main/java/com/steamclock/steamclog/Steamclog.kt:61]: If AutoRotateConfig already contains a default for fileRotationSeconds, do we need to set it here?

**Review by ssawchenko — APPROVED** [review]
Looks good, just a few non-blocking comments. If we're now saving up to 10 files we may want to look into the edge case where the device has run out of memory and cannot create a new file. Honestly I'm not sure the best way to test that, and we could defer that testing to another Issue ticket (especially since I do not believe we were handling OOM previously).

File: steamclog/src/main/java/com/steamclock/steamclog/Steamclog.kt (hunk 2 above)
Comment (jacobminer) [reply steamclog/src/main/java/com/steamclock/steamclog/Steamclog.kt]: I think we still need to set it in both places, but I've cleaned it up a bit!

File: steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt [hunk 3]
```diff
@@ -157,11 +160,51 @@ internal class ExternalLogFileDestination : Timber.DebugTree() {
...
+        val currentFile = findAvailableLogFile(expiryMs)
+        updateCachedLogFile(currentFile)
+        return currentFile
+    }
+
+    private fun isCachedLogFileValid(expiryMs: Long): Boolean {
+        // if you have a cached log file and you checked it within the expiry window
+        val now = Date().time
+        val cachedCurrentLogFile = cachedCurrentLogFile
+        val currentLogFileCachedTime = currentLogFileCachedTime
```
Comment (munl) [inline steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:178]: why are you making a copy here?

**Review by munl — APPROVED** [review]
looks good just left 1 comment

File: steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt (hunk 3 above)
Comment (jacobminer) [reply steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:178]: Because `currentLogFileCachedTime` and `cachedCurrentLogFile` are class level variables, the compiler can't smart-cast to a non-nullable version in the check below
```kotlin
return cachedCurrentLogFile != null &&
                currentLogFileCachedTime != null &&
                now - currentLogFileCachedTime < expiryMs
```

We assign a copy and check against that, since there's no way for the value to change in between the null check and the usage of the variable. We can safely remove the copy of `cachedCurrentLogFile` though, as the value isn't being used after the initial null check!

Comment (jacobminer) [reply steamclog/src/main/java/com/steamclock/steamclog/Destinations.kt:177]: @munl Removed the copy of `cachedCurrentLogFile`!
